### PR TITLE
Feat: proxy implementation

### LINKED ISSUE #222: Proxy Support
Hello there,

it would be really nice to set a HTTP or HTTPS proxy server in the step configuration, or to just pass through the already set environmental settings of the runner. 

If it is already possible and I just missed it where to set it, let me know.

### PR DESCRIPTION
**Pull request by BlackDark**
fixes #222

*Description of changes:*
Implement proxy usage feature.
See here: https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/node-configuring-proxies.html

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by guthrey-coy** [comment]
@allisaurus @paragbhingre this is a feature that many users are want to use and has been waiting for a review for 7 months... Can someone review this PR please so there can be forward progress made on this feature? Thank you!

**Comment by lcmgh** [comment]
Bump because otherwise the action is not usable in corporate environments.

**Comment by lcmgh** [comment]
@BlackDark can you provide a doc for the usage? 

Is the action now supposed to set the proxy automatically?

```
 - name: Configure AWS Credentials
      uses: BlackDark/configure-aws-credentials@580e65787125f938b804a5b3debf577dd10187a6 # aws-actions/configure-aws-credentials@v1
      with:
        aws-access-key-id: ${{ inputs.AWS_ACCESS_KEY_ID }}
        aws-secret-access-key: ${{ inputs.AWS_SECRET_ACCESS_KEY }}
        aws-region: ${{ inputs.AWS_REGION }}
        role-to-assume: ....
        http-proxy: ${{ env.HTTP_PROXY }}
```

leads to

```
Warning: Unexpected input(s) 'http-proxy', valid inputs are 
```

**Comment by lcmgh** [comment]
@BlackDark can you provide a doc for the usage? 

Is the action now supposed to set the proxy automatically?

```
 - name: Configure AWS Credentials
      uses: BlackDark/configure-aws-credentials@580e65787125f938b804a5b3debf577dd10187a6 # aws-actions/configure-aws-credentials@v1
      with:
        aws-access-key-id: ${{ inputs.AWS_ACCESS_KEY_ID }}
        aws-secret-access-key: ${{ inputs.AWS_SECRET_ACCESS_KEY }}
        aws-region: ${{ inputs.AWS_REGION }}
        role-to-assume: ....
        http-proxy: ${{ env.HTTP_PROXY }}
```

leads to

```
Warning: Unexpected input(s) 'http-proxy', valid inputs are ['aws-access-key-id', 'aws-secret-access-key', 'aws-session-token', 'aws-region', 'mask-aws-account-id', 'role-to-assume', 'web-identity-token-file', 'role-duration-seconds', 'role-session-name', 'role-external-id', 'role-skip-session-tagging']
Run BlackDark/configure-aws-credentials@580e65787125f938b804a5b3debf577dd10[18]
```

**Comment by BlackDark** [comment]
> @BlackDark can you provide a doc for the usage?
> 
> For some reason I don't get it working
> 
> ```
>  - name: Configure AWS Credentials
>       uses: BlackDark/configure-aws-credentials@580e65787125f938b804a5b3debf577dd10187a6 # aws-actions/configure-aws-credentials@v1
>       with:
>         aws-access-key-id: ${{ inputs.AWS_ACCESS_KEY_ID }}
>         aws-secret-access-key: ${{ inputs.AWS_SECRET_ACCESS_KEY }}
>         aws-region: ${{ inputs.AWS_REGION }}
>         role-to-assume: ....
>         http-proxy: ${{ env.HTTP_PROXY }}
> ```
> 
> leads to
> 
> ```
> Warning: Unexpected input(s) 'http-proxy', valid inputs are ['aws-access-key-id', 'aws-secret-access-key', 'aws-session-token', 'aws-region', 'mask-aws-account-id', 'role-to-assume', 'web-identity-token-file', 'role-duration-seconds', 'role-session-name', 'role-external-id', 'role-skip-session-tagging']
> Run BlackDark/configure-aws-credentials@580e65787125f938b804a5b3debf577dd10[18]
> ```
> 
> Probably because the `action.yml` was not adapted.

Hi @lcmgh,

rebased the PR and changed the proxy lib to something more up2date.
Also added the action.yml file but not updated the dist folder in this PR. If required for the PR to merge, will do.

If you want to test it i created another branch in my fork withe the dist updated: 

```yaml
      - name: Configure AWS Credentials
        uses: BlackDark/configure-aws-credentials@test/proxy-dist
```

**Comment by peterwoodworth** [comment]
Thanks all for your patience, I'll try to give this a review in the next week or so 🙂

**Review by peterwoodworth — CHANGES_REQUESTED** [review]
This looks good @BlackDark, thank you! I left some comments, including a couple on `http-proxy-ignore`. Before you address those however, I am not totally convinced we need this input in the first place. When would someone need to set this to true? That would imply they've set the proxy - when would someone set this and _not_ want to use it?

File: index.js [hunk 1]
```diff
@@ -259,6 +260,31 @@ const retryAndBackoff = async (fn, isRetryable, retries = 0, maxRetries = 12, ba
   }
 }
 
+function configureProxy(inputIgnore, inputProxy) {
```
Comment (peterwoodworth) [inline index.js:263]: Nitpicky, but I'm not keen on these variable names. I'd prefer they be named how they are when you pass them in

**Comment by BlackDark** [comment]
> This looks good @BlackDark, thank you! I left some comments, including a couple on `http-proxy-ignore`. Before you address those however, I am not totally convinced we need this input in the first place. When would someone need to set this to true? That would imply they've set the proxy - when would someone set this and _not_ want to use it?

Initially i though if the underlying machine has the default proxy vars set and you would want this action not use those. But maybe as you write it does not make a lot of sense to make it directly ignorable by setting a variable.

**Comment by BlackDark** [comment]
- rebased
- removed the ignore variable
- added sample to readme

**Review by peterwoodworth — CHANGES_REQUESTED** [review]
Left just a couple more comments.

Aside from the comments I left, I want to be sure that enabling this new feature by default when `http_proxy` env variable is set shouldn't break any customers out there. There is no scenario you can think of where this would be set but they wouldn't want to use this? I can't think of any reason this would be the case. Let me know what you think

**Comment by BlackDark** [comment]
> Left just a couple more comments.
> 
> Aside from the comments I left, I want to be sure that enabling this new feature by default when `http_proxy` env variable is set shouldn't break any customers out there. There is no scenario you can think of where this would be set but they wouldn't want to use this? I can't think of any reason this would be the case. Let me know what you think

I think this shouldn't break existing customers or  at least shouldn't produce unexpected result. Many other applications like for example `awscli` are also using the env if set in the environment. Or the github action runners itself also read those env vars if set: https://docs.github.com/en/actions/hosting-your-own-runners/using-a-proxy-server-with-self-hosted-runners

File: README.md [hunk 2]
```diff
@@ -245,6 +246,30 @@ with:
   web-identity-token-file: /var/run/secrets/eks.amazonaws.com/serviceaccount/token
 ```
 
+### Proxy Configuration
+
+If you run in self-hosted environments and in secured environment where you need use a specific proxy you can it in the action manually.
```
Comment (peterwoodworth) [inline README.md:251]: > If you run in self-hosted environments and in secured environment where you need use a specific proxy you can it in the action manually.

I think you missed a word here. Change to something like

> If you run in self-hosted environments and in secured environment where you need use a specific proxy you can set it in the action manually.

**Review by peterwoodworth — COMMENTED** [review]
I've tested this and it seems to work. Code also looks good. One last hangup I have is that is there any reason this is only supporting `http` and not `https`? `httpOptions` accepts an `https.Agent`, and Github Runners support https proxies as well. So it looks to me like we should be able to support `https` pretty easily.

If we can support this, I'd like to prioritize setting proxy by action input > setting proxy by env variable. And if both http and https are provided as action input, or if both are provided as env without only one being provided as an action input, then the action throws an error. Thoughts?

**Comment by BlackDark** [comment]
> I've tested this and it seems to work. Code also looks good. One last hangup I have is that is there any reason this is only supporting `http` and not `https`? `httpOptions` accepts an `https.Agent`, and Github Runners support https proxies as well. So it looks to me like we should be able to support `https` pretty easily.
> 
> If we can support this, I'd like to prioritize setting proxy by action input > setting proxy by env variable. And if both http and https are provided as action input, or if both are provided as env without only one being provided as an action input, then the action throws an error. Thoughts?

Sorry for the late reply. Have forgotten about this PR.
In the implementation we are only using an https agent: https://github.com/aws-actions/configure-aws-credentials/pull/246/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R6

Not sure what the difference would be in adding an http agent additionally. At least for all our usages it works as expected. (working with my fork since over a year now).

File: README.md (hunk 2 above)
Comment (BlackDark) [reply README.md]: done

**Comment by kellertk** [comment]
LGTM @peterwoodworth, ready to ship?

**Review by peterwoodworth — APPROVED** [review]
Thanks a ton @BlackDark for your cooperation and contribution especially after such a long wait. LGTM

**Comment by BlackDark** [comment]
> Thanks a ton @BlackDark for your cooperation and contribution especially after such a long wait. LGTM

Happy that we finally got it ready :)

**Pull request merged**